Add support for weapon, armor, and shield grade

CarlosFdez opened the pull request:
Shields have a grade but is not currently implemented. On Tikael's suggestion, we are errating kingmaker ones into starfinder ones.

CarlosFdez committed:
Add support for weapon and armor grades

CarlosFdez committed:
Add confirmation dialog when adding/remove starfinder traits

CarlosFdez commented:
A confirmation dialog has been added when adding/removing the traits from weapons and armor. The dialog only shows when the weapon/armor has runes or is non-commercial. The dialog also adds a button to swap to the opposite tech trait, in case the intent was to change from an analog to tech item or vice versa.

stwlam commented on src/module/item/physical/document.ts:
```diff
@@ -739,6 +739,55 @@ abstract class PhysicalItemPF2e<TParent extends ActorPF2e | null = ActorPF2e | n
+                const key = `PF2E.Item.Physical.ChangeEquipmentSystem.${becomingSF2e ? "ToStarfinder" : "ToPathfinder"}`;
+                const removedTrait = beforeSF2eTraits.find((t) => !newTraits.includes(t));
+                const otherTrait = removedTrait === "tech" ? "analog" : "tech";
+                const newTrait = newTraits.find((t) => t === "tech" || t === "analog");
```
```suggestion
                const newTrait = newTraits.find((t) => ["tech", "analog"].includes(t));
```
Or was there a reason for expanding this out?

CarlosFdez replied:
This one is outdated, but the reason was stricter typing (and marginally better perf doesn't hurt). I can swap it over since that doesn't matter too much.

CarlosFdez committed:
Add shield support and fix conversion confirmation when swapping between tech and analog

CarlosFdez commented:
I added shield support and fixed a bug with the conversion dialog that I caught while working on it. Shield item name generation doesn't work yet, but that's because the sf2e module doesn't (or can't) add base shield types, which is a separate thing to resolve.

stwlam commented on src/module/item/armor/document.ts:
```diff
@@ -92,9 +92,18 @@ class ArmorPF2e<TParent extends ActorPF2e | null = ActorPF2e | null> extends Phy
         const baseTraits = this.system.traits.value;
+
+        // Determine if this armor uses runes or CTASEUP and clear the opposing data
+        const isSF2e = baseTraits.some((v) => v === "tech" || v === "analog");
```
```suggestion
        const isSF2e = baseTraits.some((v) => ["tech", "analog"].includes(v));
```
It really is magical

stwlam commented on src/module/item/armor/sheet.ts:
```diff
@@ -2,9 +2,10 @@ import { AutomaticBonusProgression as ABP } from "@actor/character/automatic-bon
     CoinsPF2e,
+    type Grade,
     MATERIAL_DATA,
-    MaterialSheetData,
-    PhysicalItemSheetData,
+    type MaterialSheetData,
+    type PhysicalItemSheetData,
```
This isn't really necessary if it's a mix: esbuild just needs to know if it can drop the imported file entirely.
